File: Generate_Data/Boundary_elements.m
function dat_tab = Boundary_elements(coords_vertex, coords_vertex2, sheet, filename)

table = readtable(filename, 'TextType', 'String', 'Sheet', sheet);
table = table2array(table);
nRows = size(coords_vertex,1);
nCols = size(table,2);

edge_length = vecnorm([coords_vertex(:,1) - coords_vertex(:,3), ...
                       coords_vertex(:,2) - coords_vertex(:,4)], 2, 2);

u1 = zeros(nRows,1); u2 = zeros(nRows,1); u3 = zeros(nRows,1);
v1 = u1; v2 = u2; v3 = u3; p1 = u1; p2 = u2; p3 = u3;
res_u1 = u1; res_u2 = u2; res_u3 = u3; res_v1 = u1; res_v2 = u2; res_v3 = u3;
res_p1 = u1; res_p2 = u2; res_p3 = u3; resu1 = u1; resu2 = u2; resu3 = u3;
resv1 = u1; resv2 = u2; resv3 = u3; resp1 = u1; resp2 = u2; resp3 = u3;
Re_gem = zeros(nRows,1); CFL = Re_gem; Cx = Re_gem; Cy = Re_gem;
u_new = Re_gem; v_new = Re_gem; p_new = Re_gem; itercount = Re_gem;

for i = 1:nRows
    idx2 = coords_vertex(i,2);
    idx4 = coords_vertex(i,4);
    if idx2 < 1 || idx2 > 3 || isnan(idx2), idx2 = 1; end
    if idx4 < 1 || idx4 > 3 || isnan(idx4), idx4 = 2; end

    if i > size(table,1), break; end
    if (27+idx2) > nCols
        warning('Row %d skipped: invalid column index', i);
        continue;
    end

    u1(i) = table(i,3+idx2);
    u2(i) = table(i,3+idx4);
    v1(i) = table(i,6+idx2);
    v2(i) = table(i,6+idx4);
    p1(i) = table(i,9+idx2);
    p2(i) = table(i,9+idx4);
    res_u1(i) = table(i,12+idx2);
    res_u2(i) = table(i,12+idx4);
    res_v1(i) = table(i,15+idx2);
    res_v2(i) = table(i,15+idx4);
    res_p1(i) = table(i,18+idx2);
    res_p2(i) = table(i,18+idx4);
    resu1(i) = table(i,21+idx2);
    resu2(i) = table(i,21+idx4);
    resv1(i) = table(i,24+idx2);
    resv2(i) = table(i,24+idx4);
    resp1(i) = table(i,27+idx2);
    resp2(i) = table(i,27+idx4);
    itercount(i) = table(i,32);
end

dat_tab = [edge_length, zeros(nRows,2), u1, u2, u3, v1, v2, v3, ...
           p1, p2, p3, res_u1, res_u2, res_u3, res_v1, res_v2, res_v3, ...
           res_p1, res_p2, res_p3, resu1, resu2, resu3, resv1, resv2, resv3, ...
           resp1, resp2, resp3, Re_gem, itercount, CFL, Cx, Cy, u_new, v_new, p_new];
end

File: Compare_Results/BoundaryElements.py
import numpy as np
from mph import Model

def boundary_elements(coords_vertex2, coords_vertex, dat_tab):
    edge_length = np.linalg.norm(coords_vertex[:, 0:2] - coords_vertex[:, 2:4], axis=1)

    n = len(coords_vertex)
    dat_out = np.zeros((n, 38))  

    for i in range(n):
        cv = coords_vertex[i, :].astype(int)
        cv2 = coords_vertex2[i, :].astype(int)

        u1 = dat_tab[i, 3 + cv2[0]]
        u2 = dat_tab[i, 3 + cv2[1]]
        u3 = 0
        v1 = dat_tab[i, 6 + cv2[0]]
        v2 = dat_tab[i, 6 + cv2[1]]
        v3 = 0
        p1 = dat_tab[i, 9 + cv2[0]]
        p2 = dat_tab[i, 9 + cv2[1]]
        p3 = 0
        res_u1 = dat_tab[i, 12 + cv2[0]]
        res_u2 = dat_tab[i, 12 + cv2[1]]
        res_u3 = 0
        res_v1 = dat_tab[i, 15 + cv2[0]]
        res_v2 = dat_tab[i, 15 + cv2[1]]
        res_v3 = 0
        res_p1 = dat_tab[i, 18 + cv2[0]]
        res_p2 = dat_tab[i, 18 + cv2[1]]
        res_p3 = 0
        resu1 = dat_tab[i, 21 + cv2[0]]
        resu2 = dat_tab[i, 21 + cv2[1]]
        resu3 = 0
        resv1 = dat_tab[i, 24 + cv2[0]]
        resv2 = dat_tab[i, 24 + cv2[1]]
        resv3 = 0
        resp1 = dat_tab[i, 27 + cv2[0]]
        resp2 = dat_tab[i, 27 + cv2[1]]
        resp3 = 0

        Re_gem = 0
        itercount = dat_tab[i, 32]
        CFL = 0
        Cx = 0
        Cy = 0
        u_new = 0
        v_new = 0
        p_new = 0

        dat_out[i, :] = [
            edge_length[i], 0, 0,
            u1, u2, u3, v1, v2, v3, p1, p2, p3,
            res_u1, res_u2, res_u3, res_v1, res_v2, res_v3,
            res_p1, res_p2, res_p3, resu1, resu2, resu3,
            resv1, resv2, resv3, resp1, resp2, resp3,
            Re_gem, itercount, CFL, Cx, Cy, u_new, v_new, p_new
        ]

    return dat_out
